Documentation and test of `PointerWrapper`s (#67)
trixi-framework/P4est.jl: docs/src/introduction.md

```

[P4est.jl](https://github.com/trixi-framework/P4est.jl) wraps these C
functions accordingly and works with pointers. For a convenient alternative

> to using pointers, see the [next section](@ref pointer_wrappers). We also

Reviewer: ```suggestion
to using raw pointers directly, see the next section on [`PointerWrapper`s](@ref pointer_wrappers). We also
```
Suggestion: This gives a little more context on what to expect from the next section.

```

## [`PointerWrapper`s](@id pointer_wrappers)

> As we have seen in the [previous section](@ref translation_guidelines) many functions

Reviewer: ```suggestion
As we have seen in the previous section on [translation guidelines](@ref translation_guidelines), many functions
```
Suggestion: This way we do not assume that people read everything from top to bottom, therefore giving them some more context

provided by [`p4est`](https://github.com/cburstedde/p4est) return pointers to `struct`s,

> which requires to [`unsafe_load`](https://docs.julialang.org/en/v1/base/c/#Base.unsafe_load)
> the pointer in order to access the underlying data.

Reviewer: ```suggestion
which requires one to [`unsafe_load`](https://docs.julialang.org/en/v1/base/c/#Base.unsafe_load)
the pointer in order to access the underlying data.
```
"require" is a transitive verb and thus requires a direct object (no pun intended)

> [P4est.jl](https://github.com/trixi-framework/P4est.jl) offers a convenient way to avoid
> using `unsafe_load` whenever it is necessary by using a [`PointerWrapper`](@ref).

Reviewer: ```suggestion
[P4est.jl](https://github.com/trixi-framework/P4est.jl) offers a convenient alternative to
directly using `unsafe_load` by providing the [`PointerWrapper`](@ref) data type.
```
Suggestion: "avoid" sounds to me like it is something bad and our approach is "better"; I'd just say it's an alternative (especially we use `unsafe_load` under the hood).

> If you, e.g., have a pointer to a `p4est_connectivity` (i.e. an object of type `Ptr{p4est_connectivity}`)

Reviewer: ```suggestion
If you, e.g., have a pointer to a `p4est_connectivity` (i.e., an object of type `Ptr{p4est_connectivity}`)
```

called `connectivity`, you can use `connectivity_pw = PointerWrapper(connectivity)` to obtain
a wrapped version of the pointer, where the underlying data can be accessed simply by

> `connectivity_pw.num_trees[]` without the need to use `unsafe_load`. This works even for nested

Reviewer 2: ```suggestion
`connectivity_pw.num_trees[]` without the need to call `unsafe_load` manually. This works even for nested
```

> structures, where the data of a `struct` is, again, a `struct`. The example from above, but now

Reviewer: I don't understand the
> where the data of a `struct` is, again, a `struct`

part. Maybe you can give an example for a nested access to make it clearer?

Author: I should probably say "where a named field of a `struct` is a pointer to a `struct`", e.g., as in the struct for a `p4est`: It contains a field that is a `Ptr{p4est_connectivity_t}`. So what I want to say is that with a `PointerWrapper` it is possible to do something like:
```
connectivity = p4est_connectivity_new_brick(2, 2, 0, 0)
p4est = p4est_new_ext(MPI.COMM_WORLD, connectivity, 0, 0, true, 0, C_NULL, C_NULL)
p4est_pw = PointerWrapper(p4est)
p4est_pw.connectivity.num_trees[]
```
and one doesn't need constructions like `unsafe_load(unsafe_load(p4est).connectivity).num_trees`. I will extend the example so it also contains this case.

Author: I made the statement a little bit more precise and included a nested access in the example in [93fa7ea](https://github.com/trixi-framework/P4est.jl/pull/67/commits/93fa7ea2a0f2f7b8093a4fbae6798e3360e62b09).

> structures, where the data of a `struct` is, again, a `struct`. The example from above, but now

Reviewer: "example from above" is ambiguous; maybe this can be made more precise or with a link to it?

Author: I can give a link to the section "Translation guidelines", where the example is given, but can I also give a link to the example itself?

Reviewer: Whatever you think works best. If a direct link to the example is possible, then why not?

Author: A modified and extended the example in [93fa7ea](https://github.com/trixi-framework/P4est.jl/pull/67/commits/93fa7ea2a0f2f7b8093a4fbae6798e3360e62b09). So it is not really the same example as above. Therefore, I also don't refer to it anymore.
